20220831-add-type-restrictions-to-json-syntax.md

```diff
@@ -0,0 +1,468 @@
+   - write(user=group:2, member, group:1); invalid, the group type is not in the list of assignable relations for the member relation of the group type 
+   - write(user=user:1, parent, group:1); invalid, the user type is not in the list of assignable relations for the parent relation of the group type 
+
+2. If the user is a userset:
+   - write(user=group:2#member, member, group:1); valid, the (type=group, relation=member) is in the list of assignable relations for the member relation of the group type
+   - write(user=group:2#member, parent, group:1); invalid, the (type=group, relation=member) is not in the list of assignable relations for the parent relation of the group type
+   - write(user=group:2#parent, member, group:1); invalid, the (type=group, relation=parent) is not in the list of assignable relations for the member relation of the group type
+   - write(user=group:2#parent, parent, group:1); invalid, the (type=group, relation=parent) is not in the list of assignable relations for the parent relation of the group type
+
+3. If the user is `*`:
+   - write(user=*, parent, group:1); invalid, the parent relation on the group type does not allow for any * relation
+   - write(user=*, member, group:1); valid, the member relation on the group type allows for the * relation (and it will be considered all objects of type user)
+
+### Updating ListObjects to Respect Type Restrictions
+
+ListObjects will be the first of the Relationship Query endpoints to take advantage of this new functionality. Our hope is that we can start building a more performant ListObjects endpoint using the new type restrictions.
+
+### Updating Expand to Respect Type Restrictions
+
+Expand will need to be updated to ignore tuples that do not match the assignable tuples in the authorization model.
+
+### Updating Check to Respect Type Restrictions 
+
+THis will be the final phase and should be undertaken only once we are completely confident of how ListObjects and Expand are performing with the new functionality. Check is the core of OpenFGA, and we should be deligint in making sure it does not break and that any changes are properly communicated.
+
+When the time comes, check will be updated to ignore tuples that do not match the assignable tuples in the authorization model.
+
+# Migration
+
+## Migrating existing authorization models
+
+Existing models will not be migrated - new models will be required to use types (with an optional grace period). Authorization models that do not make use of types will not be able to use the optimized ListObjects endpoint.
+
+## Migrating existing tuples
+
+Due to the users in the tuples now required to have types, existing tuples with users as floating user_ids will no longer be valid when types are added. This will need to be communicated to users so that they can migrate accordingly. 
```

Comment:
Does it mean that only updates to these existing tuples are affected or these tuples will be "orphaned" that will have no effects in the system?

Reply:
Example of what is meant by `floating user_ids`?

Reply:
Added a definition of Users and floating user IDs to the top and clarified what happens to existing tuples with user ids lacking type info